File: app/src/main/java/org/techtown/gtguildraid/Fragments/MemberResignedFragment.java
package org.techtown.gtguildraid.Fragments;

import android.app.Dialog;
import android.graphics.Color;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.techtown.gtguildraid.Adapters.MemberCardAdapter;
import org.techtown.gtguildraid.Interfaces.MyButtonClickListener;
import org.techtown.gtguildraid.Models.GuildMember;
import org.techtown.gtguildraid.R;
import org.techtown.gtguildraid.Utils.MySwipeHelper;
import org.techtown.gtguildraid.Utils.RoomDB;

import java.util.ArrayList;
import java.util.List;

public class MemberResignedFragment extends Fragment {
    TextView resignCnt;
    RecyclerView recyclerView;
    List<GuildMember> memberList = new ArrayList<>();
    LinearLayoutManager linearLayoutManager;
    RoomDB database;
    MemberCardAdapter adapter;
    final int MAX_MEMBER = 29;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        ViewGroup view = (ViewGroup) inflater.inflate(R.layout.fragment_member_resigned, container, false);

        database = RoomDB.getInstance(getActivity());

        linearLayoutManager = new LinearLayoutManager(getActivity());
        recyclerView = view.findViewById(R.id.resignedRecyclerView);
        memberList = database.memberDao().getResignedMembers();
        recyclerView.setLayoutManager(linearLayoutManager);
        adapter = new MemberCardAdapter(memberList);
        recyclerView.setAdapter(adapter);

        resignCnt = view.findViewById(R.id.resignCnt);
        resignCnt.setText(memberList.size()+"명");

        MySwipeHelper swipeHelper = new MySwipeHelper(getActivity(), recyclerView, 200) {
            @Override
            public void instantiateMyButton(RecyclerView.ViewHolder viewHolder, List<MySwipeHelper.MyButton> buffer) {
                buffer.add(new MyButton(getActivity(),
                        "삭제",
                        50,
                        0,
                        Color.parseColor("#FF0000"),
                        new MyButtonClickListener(){
                            @Override
                            public void onClick(int pos) {
                                deleteMember(pos);
                            }
                        }));
                buffer.add(new MyButton(getActivity(),
                        "복귀",
                        50,
                        0,
                        Color.parseColor("#FFFF00"),
                        new MyButtonClickListener(){
                            @Override
                            public void onClick(int pos) {
                                recoverMember(pos);
                            }
                        }));
                buffer.add(new MyButton(getActivity(),
                        "수정",
                        50,
                        0,
                        Color.parseColor("#90ee90"),
                        new MyButtonClickListener(){
                            @Override
                            public void onClick(int pos) {
                                updateMember(pos);
                            }
                        }));
            }
        };

        return view;
    }

    private void updateMember(int pos) {
        GuildMember m = memberList.get(pos);
        final int sID = m.getID();
        String sName = m.getName();
        String sRemark = m.getRemark();

        final Dialog dialog = new Dialog(getActivity());

        dialog.setContentView(R.layout.dialog_member);

        int width = WindowManager.LayoutParams.MATCH_PARENT;
        int height = WindowManager.LayoutParams.WRAP_CONTENT;
        dialog.getWindow().setLayout(width, height);
        dialog.show();

        final EditText name = dialog.findViewById(R.id.name);
        final EditText remark = dialog.findViewById(R.id.remark);
        Button updateButton = dialog.findViewById(R.id.updateButton);

        name.setText(sName);
        remark.setText(sRemark);

        updateButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dialog.dismiss();
                String uName = name.getText().toString().trim();
                String uRemark = remark.getText().toString().trim();

                database.memberDao().update(sID, uName, uRemark);

                memberList.clear();
                memberList.addAll(database.memberDao().getResignedMembers());
                resignCnt.setText(memberList.size()+"명");

                adapter.notifyDataSetChanged();
            }
        });
    }

    private void recoverMember(int pos) {
        int memberNum = database.memberDao().getCurrentMembersWithoutMe().size();
        if(memberNum == MAX_MEMBER){
            showToast("인원이 가득찼습니다!");
            return;
        }

        GuildMember m = memberList.get(pos);
        database.memberDao().setIsResigned(m.getID(), false);

        memberList.remove(pos);
        resignCnt.setText(memberList.size()+"명");

        adapter.notifyItemRemoved(pos);
        adapter.notifyItemRangeChanged(pos, memberList.size());
    }

    private void deleteMember(int pos) {
        GuildMember m = memberList.get(pos);

        database.memberDao().delete(m);

        memberList.remove(pos);
        resignCnt.setText(memberList.size()+"명");

        adapter.notifyItemRemoved(pos);
        adapter.notifyItemRangeChanged(pos, memberList.size());
    }

    private void showToast(String msg){
        Toast.makeText(getActivity(), msg, Toast.LENGTH_LONG).show();
    }
}

File: app/src/main/java/org/techtown/gtguildraid/member/MemberResignedFragment.kt
package org.techtown.gtguildraid.member

import android.app.Dialog
import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import org.techtown.gtguildraid.R
import org.techtown.gtguildraid.common.MySwipeHelper
import org.techtown.gtguildraid.databinding.FragmentMemberResignedBinding
import org.techtown.gtguildraid.models.entities.GuildMember
import org.techtown.gtguildraid.repository.RoomDB

class MemberResignedFragment : Fragment() {
    private var _binding: FragmentMemberResignedBinding? = null
    private val binding get() = _binding!!

    var memberList: MutableList<GuildMember?>? = ArrayList()
    var database: RoomDB? = null
    private lateinit var adapter: MemberCardAdapter
    val MAX_MEMBER = 29

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentMemberResignedBinding.inflate(inflater, container, false)
        val view = binding.root

        database = RoomDB.getInstance(activity)
        memberList = database!!.memberDao().resignedMembers
        binding.resignedRecyclerView.layoutManager = LinearLayoutManager(activity)
        adapter = MemberCardAdapter(memberList)
        binding.resignedRecyclerView.adapter = adapter

        binding.resignCnt.text = getString(R.string.member_numbers_resign, memberList!!.size)

        //swipeHelper를 통해 swipe 구현
        object : MySwipeHelper(activity, binding.resignedRecyclerView, 200) {
            //instaiateMyButton 구현을 통해 button 관련 정보를 삽입
            override fun instantiateMyButton(buffer: MutableList<MyButton>) {
                buffer.add(MyButton(
                    activity,
                    "삭제",
                    50,
                    0,
                    Color.parseColor("#FF0000")
                ) { pos: Int -> deleteMember(pos) })
                buffer.add(MyButton(
                    activity,
                    "복귀",
                    50,
                    0,
                    Color.parseColor("#FFFF00")
                ) { pos: Int -> recoverMember(pos) })
                buffer.add(MyButton(
                    activity,
                    "수정",
                    50,
                    0,
                    Color.parseColor("#90ee90")
                ) { pos: Int -> updateMember(pos) })
            }
        }
        return view
    }

    private fun updateMember(pos: Int) {
        val m = memberList!![pos] ?: return
        val sID = m.id
        val sName = m.name
        val sRemark = m.remark

        val dialog = Dialog(requireActivity())
        dialog.setContentView(R.layout.dialog_member)
        val width = WindowManager.LayoutParams.MATCH_PARENT
        val height = WindowManager.LayoutParams.WRAP_CONTENT
        dialog.window!!.setLayout(width, height)
        dialog.show()
        val name = dialog.findViewById<EditText>(R.id.name)
        val remark = dialog.findViewById<EditText>(R.id.remark)
        val updateButton = dialog.findViewById<Button>(R.id.updateButton)
        name.setText(sName)
        remark.setText(sRemark)
        updateButton.setOnClickListener {
            dialog.dismiss()
            val uName = name.text.toString().trim { it <= ' ' }
            val uRemark = remark.text.toString().trim { it <= ' ' }
            database!!.memberDao().update(sID, uName, uRemark)
            memberList!!.clear()
            database!!.memberDao().resignedMembers?.let { memberList!!.addAll(it) }
            binding.resignCnt.text = getString(R.string.member_numbers_resign, memberList!!.size)
            adapter.notifyDataSetChanged()
        }
    }

    private fun recoverMember(pos: Int) {
        val memberNum = database!!.memberDao().currentMembersWithoutMe.size
        if (memberNum == MAX_MEMBER) {
            showToast()
            return
        }
        val m = memberList!![pos] ?: return
        database!!.memberDao().setIsResigned(m.id, false)
        memberList!!.removeAt(pos)
        binding.resignCnt.text = getString(R.string.member_numbers_resign, memberList!!.size)
        adapter.notifyItemRemoved(pos)
        adapter.notifyItemRangeChanged(pos, memberList!!.size)
    }

    private fun deleteMember(pos: Int) {
        val m = memberList!![pos]
        database!!.memberDao().delete(m)
        memberList!!.removeAt(pos)
        binding.resignCnt.text = getString(R.string.member_numbers_resign, memberList!!.size)
        adapter.notifyItemRemoved(pos)
        adapter.notifyItemRangeChanged(pos, memberList!!.size)
    }

    private fun showToast() {
        Toast.makeText(activity, "인원이 가득찼습니다!", Toast.LENGTH_LONG).show()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}